### PR TITLE
Flink: Add Sink options to override the compression properties of the Table

### PR DESCRIPTION
**Pull request by pvary**
Based on my performance tests using `snappy` codec to write out files is 40% more performant that writing with gzip.
Since with Flink there is an additional compaction step which overwrites the original data files anyway, it would be useful to first write out the files with `snappy` and use the table codec only at the compaction time.

To provide this functionality additional `FlinkWriteOptions` are added to modify the table default settings for:
- compression-codec
- compression-level - For Parquet and Avro formats
- compression-strategy - For ORC formats

### CODE REVIEW

**Review by nastra — APPROVED** [review]
changes LGTM and make sense, but would be good to have somebody with more Flink knowledge review this as well

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkConfParser.java [hunk 1]
```diff
@@ -139,7 +139,6 @@ public StringConfParser defaultValue(String value) {
     }
 
     public String parse() {
-      Preconditions.checkArgument(defaultValue != null, "Default value cannot be null");
```
Comment (stevenzwu) [inline flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkConfParser.java:142]: why do we remove this line? My understanding of code is that `parseOptional` should be used if default can be null

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java [hunk 2]
```diff
@@ -564,14 +571,58 @@ static IcebergStreamWriter<RowData> createStreamWriter(
...
+   * @param table The table to get the table level settings
+   * @param format The FileFormat to use
+   * @param conf The write configuration
+   * @return The properties to use for writing
+   */
+  private static Map<String, String> writeProperties(
+      Table table, FileFormat format, FlinkWriteConf conf) {
+    Map<String, String> writeProperties = Maps.newHashMap(table.properties());
+
+    String level;
```
Comment (stevenzwu) [inline flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:600]: nit: not necessary to define it outside the switch block. should we define it as local to the case block?

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java [hunk 3]
```diff
@@ -564,14 +571,58 @@ static IcebergStreamWriter<RowData> createStreamWriter(
...
+        if (level != null) {
+          writeProperties.put(PARQUET_COMPRESSION_LEVEL, level);
+        }
+
+        break;
+      case AVRO:
+        writeProperties.put(AVRO_COMPRESSION, conf.avroCompressionCodec());
+        level = conf.avroCompressionLevel();
+        if (level != null) {
+          writeProperties.put(AVRO_COMPRESSION_LEVEL, conf.avroCompressionLevel());
```
Comment (stevenzwu) [inline flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:614]: nit: can use `level` in this line.

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkConfParser.java (hunk 1 above)
Comment (pvary) [reply flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkConfParser.java:142]: The default value for `AVRO_COMPRESSION_LEVEL_DEFAULT`, `PARQUET_COMPRESSION_LEVEL_DEFAULT` is `null`, so we do not have default values there

Comment (pvary) [reply flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkConfParser.java]: I missed this... Thanks for catching!

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java (hunk 2 above)
Comment (pvary) [reply flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java]: I hate when the blocks in the `switch` are not independent.
I refactored to use a different variable instead inside each block

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java (hunk 3 above)
Comment (pvary) [reply flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:613]: I hate when the blocks in the switch are not independent.
I refactored to use a different variable instead inside each block

File: flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java (hunk 2 above)
Comment (pvary) [reply flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java]: Removed

**Review by stevenzwu — APPROVED** [review]

**Comment by stevenzwu** [comment]
Thanks @pvary for the contribution and @nastra for the review

**Comment by pvary** [comment]
Thanks @stevenzwu and @nastra for the review!

**Pull request merged**